File: code/mobile_app/lib/data/models/locker.dart
class Locker {
  const Locker({
    required this.id,
    required this.code,
    required this.isBooked,
    required this.lockState,
    required this.doorState,
    required this.securityAlertActive,
    required this.securityAlertMessage,
    this.currentUserId,
    this.activeRequestId,
    this.reservedAt,
    this.overdueReleasedAt,
  });

  final String id;
  final String code;
  final bool isBooked;
  final String lockState; // LOCKED, UNLOCKED, UNKNOWN
  final String doorState; // OPEN, CLOSED, UNKNOWN
  final bool securityAlertActive;
  final String securityAlertMessage;
  final String? currentUserId;
  final String? activeRequestId;
  final DateTime? reservedAt;
  final DateTime? overdueReleasedAt;

  factory Locker.fromJson(Map<String, dynamic> json) {
    final userVal = json['currentUserId'];
    String? uId;
    if (userVal is Map<String, dynamic>) {
      uId = userVal['_id']?.toString();
    } else {
      uId = userVal?.toString();
    }

    final requestVal = json['activeRequestId'];
    String? reqId;
    if (requestVal is Map<String, dynamic>) {
      reqId = requestVal['_id']?.toString();
    } else {
      reqId = requestVal?.toString();
    }

    DateTime? reservedAt;
    final rawReserved = json['reservedAt'];
    if (rawReserved != null) {
      reservedAt = DateTime.tryParse(rawReserved.toString());
    }

    DateTime? overdueReleasedAt;
    final rawOverdueReleased = json['overdueReleasedAt'];
    if (rawOverdueReleased != null) {
      overdueReleasedAt = DateTime.tryParse(rawOverdueReleased.toString());
    }

    return Locker(
      id: json['_id']?.toString() ?? '',
      code: json['code']?.toString() ?? '-',
      isBooked: json['isBooked'] == true,
      lockState: json['lockState']?.toString() ?? 'UNKNOWN',
      doorState: json['doorState']?.toString() ?? 'UNKNOWN',
      securityAlertActive: json['securityAlertActive'] == true,
      securityAlertMessage: json['securityAlertMessage']?.toString() ?? '',
      currentUserId: uId,
      activeRequestId: reqId,
      reservedAt: reservedAt,
      overdueReleasedAt: overdueReleasedAt,
    );
  }
}

File: smart_locker/backend/src/models/Locker.js
const mongoose = require('mongoose');
const { LockerStates, DoorStates } = require('../constants/enums');

const lockerSchema = new mongoose.Schema(
  {
    stationId: { type: mongoose.Schema.Types.ObjectId, ref: 'Station', required: true },
    code: { type: String, required: true, unique: true, uppercase: true, trim: true },
    controlTopic: { type: String, required: true },
    stateTopic: { type: String, required: true },
    doorTopic: { type: String, default: '' },
    securityTopic: { type: String, default: '' },
    lockState: { type: String, enum: Object.values(LockerStates), default: LockerStates.UNKNOWN },
    doorState: { type: String, enum: Object.values(DoorStates), default: DoorStates.UNKNOWN },
    securityAlertActive: { type: Boolean, default: false },
    securityAlertMessage: { type: String, default: '' },
    securityAlertUpdatedAt: { type: Date, default: null },
    isMaintenance: { type: Boolean, default: false },
    isBooked: { type: Boolean, default: false },
    currentUserId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    activeRequestId: { type: mongoose.Schema.Types.ObjectId, ref: 'AccessRequest', default: null },
    lastSeenAt: { type: Date, default: null },

    // Overdue tracking fields
    reservedAt: { type: Date, default: null },                                         // when booking was assigned
    overduePaymentId: { type: mongoose.Schema.Types.ObjectId, ref: 'Order', default: null }, // order for overdue charge
    overdueReleasedAt: { type: Date, default: null }                                   // when payment confirmed (grace period starts)
  },
  { timestamps: true }
);

module.exports = mongoose.model('Locker', lockerSchema);
